### DISCUSSION TITLE
Recommended way of linking local images is not working, other solutions are not perfect either

### DISCUSSION THREAD

**Discussion by warriotox** [mkdocs/mkdocs#2819] [Q&A] [answered]
Hello I am new to mkdocs, but I am trying to follow the documentation, and the recommended way of linking a local image doesn't work for me. I figured out how to change the relative path to get the image to load in both local and compiled mode, but this approach results in warning during build (running `mkdocs build).

The [documentation](https://www.mkdocs.org/user-guide/writing-your-docs/#linking-to-images-and-media) has the path start without a forward slash and being relative to the /docs/ folder.
![image](https://user-images.githubusercontent.com/5864114/162440997-6c1caa0d-a611-4f9d-ab13-640847024c1a.png)
![image](https://user-images.githubusercontent.com/5864114/162439605-e839b89e-b1f0-4c44-8dd3-36e9989ecb77.png)
But when I try this approach the image does not load. Am I missing something? And how come my approach the works results in errors during build?

I am using mkdocs version 1.3.0

my folder structure:
![image](https://user-images.githubusercontent.com/5864114/162439980-ce9b909d-ffce-4339-bbc7-5a4ccef3feba.png)

### Recommended (non-functioning) Approach
This is exactly how it should be based on the documentation.

my code: `![](best-practices/reading-order_ex1.png)`

error during build:
![image](https://user-images.githubusercontent.com/5864114/162440092-a9f7010b-8b3f-4276-b3e2-43b71bed6af5.png)

compiles to this:
![image](https://user-images.githubusercontent.com/5864114/162440215-ca3c230a-f172-419a-81fe-fe0976a50c8c.png)

throws 404 in the browser.

### Fully Functioning Approach (but with errors during build)

my code: `![](../../best-practices/reading-order_ex1.png)` But this errors during build, but works local and in compiled

error during build:
![image](https://user-images.githubusercontent.com/5864114/162439468-4a99cbdb-6bbd-4584-896c-7d6e80b04b22.png)

compiles to this:
![image](https://user-images.githubusercontent.com/5864114/162439363-096ae1c3-8994-45db-9d92-1f909a1edce4.png)

### Local Functioning Approach

my code: `![](/best-practices/reading-order_ex1.png)` no errors during build, but works only local

compiles to this:
![image](https://user-images.githubusercontent.com/5864114/162439074-f1249f77-1595-44f7-9e25-8f723c0670e7.png)

throws 404 in the browser:
![image](https://user-images.githubusercontent.com/5864114/162439188-57656735-5039-47bd-980f-81bd3f6223cb.png)

**Comment by oprypin** [comment, marked as answer]
The paths are relative to the current document (the path it's in).

`![](reading-order_ex1.png)` is correct if you're currently in best-practices/reading-order.md.

Paths starting with slash are indeed unsupported.

**Reply by warriotox** [reply]
Hi @oprypin, but that is exactly my problem. Look at the folder structure image I attached. The reading-order.md file is inside the best-practices folder, and the image is right there next to it. Yet the image isn't rendered on the page and throws 404 Instead. Only when I added `../../` infront of the file name then it works on local and in built version. And when I add only `/` infront of the file name it works only on local. That is why I wrote the post, to find out why the recommended approach is failing for me...

**Reply by pawamoy** [reply]
> The reading-order.md file is inside the best-practices folder, and the image is right there next to it

Yes, that's why you need to use `![](reading-order_ex1.png)` and not `![](best-practices/reading-order_ex1.png)`.
It's really like folders: if you're in folder `/A` (or your current opened file is in folder `/A`), to print file `/A/1.png` you do `cat 1.png`, not `cat A/1.png`.

Not saying the documentation is crystal-clear about this though.

**Reply by warriotox** [reply]
Oh my ***! I am blind and stupid. :X Thank you both and apologies for wasting your time. I see it now...

**Comment by warriotox** [comment]


**Reply by warriotox** [reply]
@pawamoy @oprypin  guys please can you check the above? I am starting to loose it...

**Reply by Andre601** [reply]
Did you try something like `![image](./image.png)`?

`.` tells the browser to start from the same directory/page as the html file.

**Reply by warriotox** [reply]
Everything works fine. I am just no with it 😭 The reason it didn't work for me in production is because I didn't commit the images to git which is how they get to the server... 🤦

**Reply by vishalgakhare** [reply]
Hi @warriotox - I'm facing the same issue. Since I have all my images in a different folder, this is how the markdown looks like

```md
![ISRM table](../../_attachments/etb1293.png)
```

The image renders locally in markdown ...proving the relative path is correct...but not when I build the mkdocs site.
I did inspect the element and it looks like the path is not changed.

```html
<img src="../../_attachments/etb1293.png" alt="ISRM table">
```

**Reply by mjbear** [reply]
> The image renders locally in markdown ...proving the relative path is correct...but not when I build the mkdocs site. I did inspect the element and it looks like the path is not changed.
> 
> ```
> <img src="../../_attachments/etb1293.png" alt="ISRM table">
> ```

@vishalgakhare
By chance did you solve your relative image linking issue?

I'm also working on a situation where the image is nested a few levels deep and relative links throw errors with mkdocs.

_(Though it would seem our scenarios are different than the OP since it sounds like their image is in the same directory as the Markdown file they are displaying it in.)_

**Update:** By the [mkdocs documentation for links and images](https://mkdocs.readthedocs.io/en/0.10/user-guide/writing-your-docs/#linking-documents) that we're creating relative URLs as described.

**Reply by kamilkrzyskow** [reply]
Read up on this setting:
- https://www.mkdocs.org/user-guide/configuration/#use_directory_urls

And understand what it means to use `../`. It means to go back up by 1 directory level.
When using the Markdown-based links you're working with the filesystem links, so if a file is 2 directories above then you use 
```md
![](../../image.png)
```
However, when using HTML you're referring to the URL in your browser. By default you're using the directory urls, so it adds one `/` at the end, which adds 1 more directory level you have to go back.

So raw HTML tags need to take this into account and add 1 more `../`

```html
<img src="../../../image.png" alt="">
```

**Reply by mjbear** [reply]
> However, when using HTML you're referring to the URL in your browser. By default you're using the directory urls, so it adds one `/` at the end, which adds 1 more directory level you have to go back.
> 
> So raw HTML tags need to take this into account and add 1 more `../`
> 
> ```
> <img src="../../../image.png" alt="">
> ```

@kamilkrzyskow
The Markdown renders properly on GH with just two `../..` and adding a third one would break the link.

I'm not doubting what you may have found.
Did you test your theory against mkdocs (as in build some docs)?
Thank you!

**Reply by kamilkrzyskow** [reply]
> Did you test your theory against mkdocs (as in build some docs)?

Sure, I work with a lot of MkDocs related projects. Here is an example:
- [2819-mkdocs.zip](https://github.com/user-attachments/files/19394439/2819-mkdocs.zip)

<details>
<summary> mkdocs serve </summary>

![image](https://github.com/user-attachments/assets/d149d95a-bb0b-40cd-9d4f-2c249e0bfc39)

</details>

**Reply by mjbear** [reply]
@kamilkrzyskow
Thank you.
In your example, I see the Markdown has two `../` whereas the inline HTML has three.

I'm only trying to use Markdown.
The image I have is in a sibling directory at roughly the same level (no the same directory).

So the Markdown path for the image to display in `docs/lab1/README.md` should be something such as `../../labs/lab1/image.png` ... right?

```
.
├── docs
│   └── lab1
│       └── README.md
├── labs
│   └── lab1
│       ├── image.png
│       └── README.md
```

**Reply by kamilkrzyskow** [reply]
The `labs` directory is outside of the `docs` directory. MkDocs only processes files inside the [`docs`](https://www.mkdocs.org/user-guide/configuration/#docs_dir) directory, and adding even 10x `../` in your situation will still be "capped" by the root `docs` directory, and therefore the image can't be reached. I think this case should be covered with a proper warning 🤔 maybe increase the severity of https://www.mkdocs.org/user-guide/configuration/#validation

There is also the https://www.mkdocs.org/user-guide/configuration/#custom_dir setting, which is used for template overrides, and its contents are merged with the final build. This sort of allows to put images in another directory, however the root of the `custom_dir` is merged with the root of the `docs` directory, so all relative links inside the markdown pages would still have to point to the relative path, as if they would be inside the `docs` directory, not outside, so not ideal if you want to have preview inside your code editor.

If you want preview inside your code editor and after MkDocs build, don't use `custom_dir`, but instead keep every file under the `docs` file tree.
